pkg/asset/agent/image/ignition.go

```diff
@@ -120,6 +120,10 @@ func (a *Ignition) Generate(dependencies asset.Parents) error {
 		},
 	}
 
+	if len(agentManifests.NMStateConfigs) == 0 {
```

Comment:
I'd recommend to perform this check in the related asset, here: https://github.com/openshift/installer/blob/c70b4faafbe04f1d37d37465bb687c3204f34fd6/pkg/asset/agent/manifests/nmstateconfig.go#L193

Reply:
It's legal to e.g. use an empty agent-config to create the cluster-manifests and only fill in the NMStateConfigs after. So we don't want to validate too early, only when you go to create the image and you don't have any NMStateConfigs.

Reply:
Should the message be more clear that either provide at least one nmstateconfig in cluster-manifests/nmstateconfig.yaml or if providing agent-config.yaml provide at least one nmstateconfig in the networkConfig?

Reply:
That's right, but it's just the NMStateConfig asset that loads the content of the `cluster-manifests/nmstateconfig.yaml` file if present on the disk. The AgentManifests it's just a collector, and it's resolved as a dependency in stack by the Ignition one, so not sure if I'm missing something but the validation location doesn't seem the right one?

Reply:
![Screenshot from 2022-08-18 16-54-32](https://user-images.githubusercontent.com/60063538/185427328-1acf536a-4174-481e-bad8-0aa52061ee94.png)

Reply:
Btw, if none of them (agent-config and nmstateconfig) are provided I'd expect that this [validation](https://github.com/openshift/installer/blob/c70b4faafbe04f1d37d37465bb687c3204f34fd6/pkg/asset/agent/manifests/nmstateconfig.go#L190-L192) will fail

Reply:
I think what Zane has currently implemented allows a user to first run "create cluster-manifests", edit the nmstateconfig to make any changes, and then run "create image".

If the validation is placed in nmstateconfig, then "create cluster-manifests" will immediately fail, thus forcing you to provide the nmstateconfig in the agent-config.

Reply:
> Btw, if none of them (agent-config and nmstateconfig) are provided I'd expect that this validation will fail

From my testing, the validation in ignition works.

Reply:
> I think what Zane has currently implemented allows a user to first run "create cluster-manifests", edit the nmstateconfig to make any changes, and then run "create image".
> 
> If the validation is placed in nmstateconfig, then "create cluster-manifests" will immediately fail, thus forcing you to provide the nmstateconfig in the agent-config.

Bingo :+1:

Reply:
I'm not sure if the `create cluster-manifests` it's in good shape now, but if you run it with just the install-config.yaml (and without an agent-config.yaml, so that it will get generated, currently the [nil check ](https://github.com/openshift/installer/blob/c70b4faafbe04f1d37d37465bb687c3204f34fd6/pkg/asset/agent/manifests/nmstateconfig.go#L190-L192) will already fail

Reply:
| f the validation is placed in nmstateconfig, then "create cluster-manifests" will immediately fail, thus forcing you to provide the nmstateconfig in the agent-config.

So wouldn't be better in general to provide an agent-config template so that the NMStateConfig will not fail, rather than moving the check into another asset?

Reply:
> So wouldn't be better in general

This check is going away again in a couple of weeks, as soon as we have fixed all of the linked bugs. Let's not get _too_ general.

> to provide an agent-config template so that the NMStateConfig will not fail, rather than moving the check into another asset?

That wouldn't solve the case where you intentionally create an agent-config with no NMState data, with the intent to customise the manifests to add NMState data after `create cluster-manifests`.
The only way to catch that case is to make sure the validation doesn't run until you do `create image`, and the only way to do that is to put it in an asset that is only required by the image target.

Reply:
Agree, especially considering if that check will go away soon. In general my concerns are about not making an asset self-sufficient, which could indicate a potential problem somewhere else

Reply:
Btw, the aforementioned scenario was still broken for me, I will re-test it with the latest changes as a double-check